File: 78.subsets.py
#
# @lc app=leetcode id=78 lang=python3
#
# [78] Subsets
#

# @lc code=start
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        path = []

        def dfs(nums, pos):
            ans.append(list(path))
            if pos == len(nums):
                return
            for i in range(pos, len(nums)):
                path.append(nums[i])
                dfs(nums, i+1)
                path.pop()
        
        dfs(nums, 0)
        return ans
        
# @lc code=end

File: 78.subsets.cpp
/*
 * @lc app=leetcode id=78 lang=cpp
 *
 * [78] Subsets
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> path;
        vector<vector<int>> ans;

        dfs(nums, 0, path, ans);
        return ans;
    }

    void dfs(vector<int>& nums, int pos, vector<int>& path, vector<vector<int>>& ans) {
        ans.push_back(path);
        if(pos == nums.size()) return;

        for(int i=pos; i<nums.size(); ++i) {
            path.push_back(nums[i]);
            dfs(nums, i+1, path, ans);
            path.pop_back();
        }
    }
};
// @lc code=end

